src/gui/presets.h

```diff
@@ -30,6 +30,31 @@ typedef enum dt_gui_presets_format_flag_t
   FOR_NOT_COLOR = 1 << 4
 } dt_gui_presets_format_flag_t;
 
+typedef struct dt_gui_presets_edit_dialog_t
```

Comment:
Feels like this should be a private structure to presets.c? I assume no other code should change anything here?

Reply:
no, this struct is passed to the final callback function that is used by the dialog requester to do whatever they need (usually refresh)

Reply:
Then you could provide a refresh function? At the moment only the treeview is accessed, via an anonymous gpointer. So all the internal details of the dialog are exported (meaning possibly it becomes hard to change anything there later because people may have started to unpredictably access any of the data/widgets) _just_ so that an anonymous pointer can be accessed. Why not just export that? Or even the underlying treestore directly rather than the treeview, since only the treestore needs to be refreshed. It might seem like a good idea to give access to all details so people can do whatever they want, but that usually just ends up an unstructured mess. Or do you have concrete ideas of how the dialog could be extended/manipulated?

The function tree_insert_presets might be a good candidate to refactor and move to presets.c. Logically, it would make sense to have all the presets related details, dialogs and database accesses reside in that file, rather then spreading deeply into preferences.

And you are right, my preset related changes in preferences are fairly limited (and not complete yet; I was planning on doing some minor refactoring myself to avoid duplication with the last few changes).

Reply:
not sure we are speaking of the same thing... What I have done to deduplicate is to provide a "public" function to show the "edit presets" dialog.
This function is used in 3 places now : 
- the "edit preset" menuitem in the presets menu of modules
- the double-click on a preset line in the preference dialog
- in the modulegroups preset manager

so we don't to rely on treeview/treestore or whatever : we need a more general thing.
That said, you're right that we don't need to expose the full structure inside the final callback, but sadly we need to expose more than "just" the input data : in some cases, we need to know what has been changed. So the simplest was to expose the entire structure... (Note that this struct is only passed inside the final callback which is use after the dialog has been "validated")